src/databricks/labs/remorph/reconcile/query_builder/threshold_query.py

```diff
@@ -0,0 +1,170 @@
+)
+from databricks.labs.remorph.reconcile.recon_config import Thresholds
+from databricks.labs.remorph.snow.databricks import Databricks
+
+logger = logging.getLogger(__name__)
+
+
+class ThresholdQueryBuilder(QueryBuilder):
+    # Comparison query
+    def build_comparison_query(self) -> str:
+        select_clause, where = self._generate_select_where_clause()
+        from_clause, join_clause = self._generate_from_and_join_clause()
+        # for threshold comparison query the dialect is always Daabricks
+        query = select(*select_clause).from_(from_clause).join(join_clause).where(where).sql(dialect=Databricks)
+        logger.info(f"Threshold Comparison query: {query}")
+        return query
+
+    def _generate_select_where_clause(self) -> tuple[list[exp.Alias], exp.Or]:
+        thresholds = self.table_conf.thresholds
+        select_clause = []
+        where_clause = []
+
+        # threshold columns
+        for threshold in thresholds:
+            column = threshold.column_name
+            base = exp.Paren(
+                this=build_sub(
+                    left_column_name=column,
+                    left_table_name="source",
+                    right_column_name=column,
+                    right_table_name="databricks",
+                )
+            ).transform(coalesce)
+
+            select_exp, where = self._build_expression_type(threshold, base)
+            select_clause.extend(select_exp)
+            where_clause.append(where)
+        # join columns
+        for column in sorted(self.table_conf.get_join_columns("source")):
+            select_clause.append(build_column(this=column, alias=f"{column}_source", table_name="source"))
+        where = build_where_clause(where_clause)
+
+        return select_clause, where
+
+    def _build_expression_alias_components(
+        self, threshold: Thresholds, base: exp.Expression
+    ) -> tuple[list[exp.Alias], exp.Expression]:
+        select_clause = []
+        column = threshold.column_name
+        select_clause.append(
+            build_column(this=column, alias=f"{column}_source", table_name="source").transform(coalesce)
+        )
+        select_clause.append(
+            build_column(this=column, alias=f"{column}_databricks", table_name="databricks").transform(coalesce)
+        )
+        where_clause = exp.NEQ(this=base, expression=exp.Literal(this="0", is_string=False))
+        return select_clause, where_clause
+
+    def _build_expression_type(
+        self, threshold: Thresholds, base: exp.Expression
+    ) -> tuple[list[exp.Alias], exp.Expression]:
+        column = threshold.column_name
+        # default expressions
+        select_clause, where_clause = self._build_expression_alias_components(threshold, base)
+
+        if (
+            threshold.get_type() == ThresholdMode.NUMBER_ABSOLUTE.value
+            or threshold.get_type() == ThresholdMode.DATETIME.value
+        ):
+            if threshold.get_type() == ThresholdMode.DATETIME.value:
```

Comment:
can this be simplified by moving this check outside

Reply:
No, Beuase we have to apply anonymous function on top of default or vice-versa.